Add SGD Optimizer with Tests

ravikiran0606 opened the pull request:
* Implement Base Class VOptimizer

* Implement Class TSGD

* Add Unit Tests for TSGD

* Add Parsing Options for Optimizers

* Modify MethodDL to include the optimizer TSGD

* Add Momentum Update method  for SGD Optimizer

* Unit Tests and MethodDL Tests are passing.

SimeonEhrig committed:
Add CUDA device compiler, which allows to generate CUDA PTX Code on runtime.

The class IncrementalCUDADeviceCompiler use external tools to generate PTX and cuda fatbin files. It runs the tools clang and fatbinary via llvm::sys::ExecuteAndWait. The class also handle to include new code in existing code. The steps of the compiler pipeline are:
- clang: CUDA C++ + previous PCH -> PCH
- clang: PCH -> PTX
- fatbinary: PTX -> fatbin

There is no selection of code. Every input of the cling will pass to the  IncrementalCUDADeviceCompiler.

SimeonEhrig committed:
CUDA device compiler can use include paths from cling runtime.

This new function allows the CUDA device compiler to search after headers in folders, which was declared at runtime vie .I command.

SimeonEhrig committed:
Overwork of the include path handling of the CUDA device compiler.

I replaced copies of the include paths with a pointer to the headerSearchOptions. Now, explicit handling of the include paths is not more necessary. Add include paths, which was declared via argument at start also works.

SimeonEhrig committed:
Setting the arguments of cling to clang nvptx and fatbinary are possible.

Now, it is possible to set some arguments of the clang nvptx and fatbinary via arguments at cling start. The arguments are filtered. So not every argument is possible at the moment. The Arguments can’t changed during runtime, because the PCH-files forbid it. For Example, the calng nvptx use the optimization level, which is set at start of cling.

At the moment, the debug options of clang nvptx are simple. If any debug option is detected, just a -g will add to the clang nvptx.

Additional PTX options for clang nvptx doesn’t works at the moment. There is a problem at parsing at the start of cling.

SimeonEhrig committed:
Add workaround for clang PCH-bug

At the moment, to extend the AST-tree of the device code, we use PCH-files to extend the exist device code with new lines of code. In detail, if we want to create a new PTX-file, we use the CUDA code (.cu file) and a PCH-file with the existing AST as input and generate an new PCH-file, which contains the whole AST. Then, the PCH-file will compiled to a PTX-file.

A bug in clang prevent, that we can’t generate more than 5 new PCH-files. The bug is not easy to fix, so I write a small workaround. Instead using a PCH-file, which contains the AST, we generate a new complete AST from all .cu-files every time.

The workaround is temporary and should removed, if clang is patched.

SimeonEhrig committed:
The CUDA compiler can handle variable declarations from the prompt.

Now, it is possible to declare variables on the prompt, which are visible for other statements.

The problems was, that cling wrapped all statements in a function, to get valid input. So, every variable is just visible inside in its own wrapper function. To solve the problem, cling change the local variable declaration to a global declaration.

The implementation for the CUDA compiler checks, if the unwrapping is happened. If it happened, the c++ code of the unwrapped variable declaration (AST-printer) instead the raw input will be written to the .cu-file.

SimeonEhrig committed:
Improve Error Message.

amadio committed:
Remove callback functionality from TBufferMerger

The callback mechanism as implemented is not what is needed by CMS,
as previously thought. In order to allow the final goal of avoiding
oversubscription of the machines on which the TBufferMerger is used
within a framework, this functionality is better removed in order to
allow doing away also with the separate output thread currently needed
by TBufferMerger.

amadio committed:
Restructure TBufferMerger to not need a separate output thread (v2)

SimeonEhrig committed:
Fix temp-path bug.

In some cases, the path of the cling temp folder contains some non printable chars at the end.
Change the handling of the path string, to solve this problem.

SimeonEhrig committed:
Improve the search of the clang instance for CUDA.

Before, it was not possible to find the clang++, which is contained in the cling, if we don't start the clang from the bin folder ('./cling -xcuda'). Now, for example it is possible to start the cling with 'bin/cling -xcuda' .

Fix a Bug, which avoid to start './cling -xcuda -fsyntax-only'.

SimeonEhrig committed:
Add test cases for CUDA features.

- CUDA __constant__ memory
- CUDA global __device__ memory
- CUDA __host__ prefix
- CUDA kernel launch with arguments
- CUDA templated kernels
- CUDA shared memory with dynamic runtime
- CUDA Streams
- test if CUDA device is available

SimeonEhrig committed:
Overwork CUDA device tests.

- add cudaDeviceSynchronize() at every kernel launch
- remove small address bug at cudaMemcpy, if host array is used
- in parallel test cases, replace fixes thread number with variable
- overworked shared memory kernel

SimeonEhrig committed:
Add some complex cuda template kernel test cases.

ashlaban committed:
[TMVA] ROOT-9081 -- Replace GradBoost with BoostType=Grad

Since some time back the gradient boosting option of TMVA is called
"BoostType=Grad", not "GradBoost". This updated textual output to use
the new name.

SimeonEhrig committed:
Add myself to the credits.

SimeonEhrig committed:
Improve CUDA device code testcases.

steremma committed:
CUDA implementation of Im2Col

steremma committed:
Refactored Im2Col tests - now taking advantage of templates to eliminate code duplication

steremma committed:
Added CUDA specialization for the existing test suite

steremma committed:
Include CUDA tests into CMake

SimeonEhrig committed:
Add function, to save faulty .cu files of CUDA device compiler.

Before this commit, the CUDA device code Compiler class overwrited source code .cu files, if the translation get wrong. Now, it renames the faulty files, so you can check, which source code cause compiler errors.

steremma committed:
Added more documentation as per code review comments

Axel-Naumann committed:
Outline ~IPruneTool() to pin vtable.

steremma committed:
Final documentation changes in accordance with the official ROOT convention guide

steremma committed:
fix a typo

steremma committed:
Tested implementation of `Flatten` on the GPU

steremma committed:
refactored tests to remove code duplications when running the same test on different architectures

steremma committed:
Added documentation for the new functionality including docstrings and in-method comments

steremma committed:
Added the CUDA implementation for Deflatteting. Added tests for both architectures to prove correctness

steremma committed:
Simplified `ReshapeLayer` API by removing redundant constructor arguments

steremma committed:
fix indexing bug in the Reference and Cpu `Reshape` implementations

steremma committed:
Added CUDA implementation for the `Reshape` method

steremma committed:
Renamed all tests to reflect the fact that they test every method of the `ReshapeLayer` class

steremma committed:
Added a test-case for the reshape method

steremma committed:
Revert "Simplified `ReshapeLayer` API by removing redundant constructor arguments"
This change needs more thought

This reverts commit 48fc44241054a0619125f70557278b1517b31436.

steremma committed:
Fixed all tests to return status code - now CI will catch broken tests. Added tests for the CPU architecture

steremma committed:
Fixed tests, now returning error code in case of failure to enable CI checks

steremma committed:
Fixed an indexing bug

SimeonEhrig committed:
Improvements for Pull Request #240

- little changes at comments and code style
- try to use const in IncrementalCUDADeviceCompiler, where is possible
- move CUDA device code compiler instance to IncrementalParser
- change the members of CuArgs to const and adjust the setCuArgs method
- use std::vector<string> instead llvm::Smallvector<const char *> to build argv for executeAndWait
- improve the error messages of generatePCH(), generatePTX() and generateFatbinary()
- replace m_Counter with a copy in IncrementalCUDADeviceCompiler to avoid involuntary changes

steremma committed:
Added CUDA implementation for Downsample

steremma committed:
Added tests for the CUDA case and refactored the testing suite to eliminate code duplication

steremma committed:
Added new testing entries to CMakeLists under the CUDA_ENABLED condition

amadio committed:
Add option to build only experimental PyROOT

amadio committed:
Import cppyy-backend into ROOT

amadio committed:
Import CPyCppyy into ROOT

amadio committed:
Import cppyy into ROOT

amadio committed:
Integrate CpyCppyy with ROOT's build system

amadio committed:
Add CMakeLists.txt for cppyy-backend

amadio committed:
Add CMakeLists.txt for cppyy projects

etejedor committed:
First skeleton for experimental PyROOT

Includes a new mechanism for adding pythonizations and two examples
of pythonizations for TTree that have been ported on top of the
new Cppyy API (one in Python, one in C++).

amadio committed:
Add CMakeLists.txt for pyroot_experimental

amadio committed:
Add CMakeLists.txt for cppyy

amadio committed:
Require cxx14=ON or cxx17=ON when building with upstream cppyy

steremma committed:
Added unit tests for back-propagation, running on all available architectures.

The tests include overlapping local views, as well as the case where depth > 0.

steremma committed:
Unified the worker method API calls to make sure all implementation expect the same arguments

steremma committed:
Fixed a bug on back-propagation detected using my unit tests

steremma committed:
Added CUDA implementation for the back-propagation

SimeonEhrig committed:
Change clang nvptx input code selection.

As input of clang nvptx there are two possible sources. The raw input of `IncrementalParser::parseInternal()` or the AST-Printer of the `llvm::module`. Now, to decide, which input source should be use, I use the value `Transaction::ConsumerCallInfo::kCCIHandleTopLevelDecl`. This allows more C++ constructs to use.

Fixed a small bug. The AST-printer prints `setValueNoAlloc()` without the scope `cling::runtime::internal`. This caused an error, because clang could not find the declaration in the global space. The solution are dummy declaration in the global space.

At the moment, templates doesn't works.

eguiraud committed:
[DF][NFC] Remove redundant override

See e.g. C++ core guidelines C.128

dpiparo committed:
[DF] Add TColumnValue extern templates for common types

this speeds up considerably the instantiation of large templates,
such as Snapshot of many columns.
It is helpful in general too.

eguiraud committed:
[DF][NFC] Add comment on the ownership of the output trees

eguiraud committed:
[DF] Test Snapshot does not write out empty tree buffers

eguiraud committed:
[DF] Give SnapshotHelperMT ownership of its output trees

We used to heap-allocate one TTree per TBB task and never deallocate them,
for fear of interfering with ROOT's garbage collection.
At the same time, though, we deregistered said trees from ROOT's gc
mechanism with `ResetBit(kMustCleanup)`. As a result I'm not sure the output
trees were ever deleted.

This commit lets SnapshotHelperMT take ownership of the output trees so
that when a new task comes along the previous output tree in use by that
slot is deleted. This has the desired side-effect of deregistering it
from the output directory (the TBufferMergerFile) to which it belongs.
This is desirable because we call `Write` on these output files, which
in turn would call `Write` on the current output tree __and all previous
output trees__ if they were still registered with the directory -- this
is not a problem in terms of final output (`Write` on old output trees
is a no-op as there is no pending buffer to flush) but is wasteful.

The deregistration of obsolete per-task output trees from the thread-local
output directories also sets the stage for fixing ROOT-9471.

Finally, output trees are re-added to the (now thread-safe) list of cleanups
to make sure that output trees are correctly deregistered from the input
tree at their destruction, lest the input tree tries to access the deleted
output trees in its dtor, as they are registered in its list of clones.

eguiraud committed:
[DF] Do not flush buffers with empty trees from MT Snapshot

With very few entries passing a given RDataFrame cutflow, it can happen
that a TBB task ends up processing zero "good" entries.
In that scenario, it could happen that the first TTree header flushed to
the output file from a Snapshot action was for a tree with no entries and
no branches. This effectively set the number of branches to
zero for all the content written to file, even if following buffers
contained well-formed TTrees.

With this commit, if the output tree contains zero entries, we avoid
flushing it to file. And since the next task will remove this TTree from
the output directory, we can be sure that the buffer with the malformed
TTree will never be written out. This fixes ROOT-9471.

yamaguchi1024 committed:
[cxxmodules] Evaluate code with interpreter

This is an alternative implementation of #2194. Changed Pythonize to
call Evaluate instead of doing Calc on code containing
cling::printValue.

stwunsch committed:
[PyROOT] Add tutorial for pretty printing feature

SimeonEhrig committed:
Added function to detect c++ attributes at function definition.

Now, it is possible to define functions with c++ attributes without the .rawInput mode. For example functions like `[[ noreturn ]] foo() { ... }` or `[[deprecated]] [[nodiscard]] int bar(){ … }`.

Axel-Naumann committed:
Rip out ruby, unmaintained and not building since years (ROOT-9193).

Axel-Naumann committed:
[NFC] Typo in comment.

Axel-Naumann committed:
Remove unmaintained and not buildable graph2d/ios.

Axel-Naumann committed:
Remove unmaintained io/chirp.

Axel-Naumann committed:
CMake: rip out chirp from build system.

pcanal committed:
Fix for https://github.com/cms-sw/cmssw/pull/22594

Instead of assigning rule that do not correspond to any specific branches
(for example rules setting transients or whole objects) to an hypothetical
'previous' branch, we switch to assign those orphan rules to either the
non-data bearing branches (split single objects or bases) or the collection
node (since the splitting completely flatten the hiearchy in this case).

This requires enhancing the list of IDs from a single list of element index
in the 'sole' current StreamerInfo to a nested list of elements that carries
along the sub-StreamerInfo information.

pcanal committed:
Avoid double write

dpiparo committed:
[IMT] Better documentation.

ellert committed:
Include test directory using ROOT_ADD_TEST_SUBDIRECTORY

Axel-Naumann committed:
[R__LOCKGUARD] Add parens around mutex expression (ROOT-9479).

Axel-Naumann committed:
Mention removal of ruby bindings.

couet committed:
Swap bits 20 and 21 to bit coherent with TGraph.h where bit 20 is kIsHighlight

This change was proposed by Jan Musinsky (the author of the Highlight mechanism)
and approved by Lorenzo.

couet committed:
Revert "Swap bits 20 and 21 to bit coherent with TGraph.h where bit 20 is kIsHighlight"

This reverts commit b9ef7b69970688dffbce1e7a4c27dcf8808db84d.

steremma committed:
Fix zero initialization in `TCudaMatrix` (#2208)

* Fixes a previously fatal runtime error when trying to initialize a cudaMatrix with zeroes.

* Fixes a previously fatal runtime error when trying to initialize a cudaMatrix with zeroes.

* Use `cudaMemset` to avoid host memory allocation

SimeonEhrig committed:
Fix cuda template specialization bug.

Problem:
If a c++ statement with a concrete template parameter will type in, cling pass the (unwrapped) input line and the specialization of the statement, generated by compilerinstance to the clang nvptx. This causes an  explicit-specialization-after-instantiation-error.

Solution:
Template specialization declaration will be save two times at a transaction. Once with the type kCCIHandleCXXImplicitFunctionInstantiation and once with the type kCCIHandleTopLevelDecl. To avoid sending a template specialization to the clang nvptx and causing an explicit-specialization-after-instantiation-error it have to check, which kCCIHandleTopLevelDecl declaration is also a kCCIHandleCXXImplicitFunctionInstantiation declaration.

SimeonEhrig committed:
Reformated the source code with clang-format.

etejedor committed:
Update PyROOT/Cppyy contributions

etejedor committed:
State which is the corresponding #if

SimeonEhrig committed:
Forgot header for clang build.

eguiraud committed:
[DF] Avoid concurrent destruction of input and output trees

At the beginning of a task a new input TChain replaced the old input
TChain for that _thread_ in TTreeProcessorMT, causing a destruction of
the previous _thread-local_ TTree.
At the same time, the initialization of a new task _in a different
thread_ could delete the output TTree of a multi-thread Snapshot
action, which was listed in the fClones of the input TTree.

This race caused the input ~TTree to access an output tree just
after its deletion, in TTree::CopyAddresses.

The root cause of this issue is that TTreeProcessorMT uses thread-local
variables while RDataFrame uses processing slots which threads can
exchange between themselves.

eguiraud committed:
[DF] Rename action helpers' InitSlot to InitTask

In line with the naming we used for the interface required by
`RInterface::Book` for user-defined action helpers.

There are several other places where `InitSlot` is used to indicate
a function that is executed to initialize a slot at the beginning
of a task. Those might be renamed as well, since `InitTask` makes
it much more clear that the function is executed multiple times
per slot.

etejedor committed:
Make AddToGlobalScope static but not inline

eguiraud committed:
[DF] Make all action helpers inherit from RActionImpl

This is in line with what we require of user-defined action helpers
in Book.

eguiraud committed:
[DF] Decorate every action helper with an empty FinalizeTask method

ashlaban committed:
ROOT-9483 -- Prevent nullptr deref. in compiled macros (#2214)

When compiling TMVA GUI methods stand alone (e.g. clang++ not through
`root -l`) the graphics subsystem of ROOT is not initialised leading
to a null pointer dereference.

eguiraud committed:
[DF] Call ActionHelper::FinalizeTask at the end of each task

Axel-Naumann committed:
Revert "Outline ~IPruneTool() to pin vtable."

This reverts commit d1cea393b3334657c010df32ca0412441898a87a.
Triggers failure of roottest/root/meta/cmsUnload that we need to debug first.

etejedor committed:
Rename init function of PyROOT

Eliminate previous redundancy with namespace.

pcanal committed:
Use same normalization for branch, element name and later searches.

Axel-Naumann committed:
Revert "Mark unimplemented new() "throw()" as they return nullptr."

This reverts commit e3b505efe73ad6f98be501764a19d5f326fb10f8.
The commit caused the compilation error:
core/newdelete/src/NewDelete.cxx:214:7: error: exception specification in declaration does not match previous declaration
void *operator new(size_t /*size*/, std::align_val_t /*al*/) throw()
      ^
/cvmfs/sft.cern.ch/lcg/contrib/gcc/7binutils/x86_64-centos7/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../include/c++/7.3.0/new:143:7: note: previous declaration is here
void* operator new(std::size_t, std::align_val_t)
      ^

etejedor committed:
Remove unnecessary includes

etejedor committed:
Remove unnecessary inline

etejedor committed:
Follow ROOT naming convention

etejedor committed:
Const-correctness fixes

etejedor committed:
Give more meaningful name to variable

lmoneta committed:
Re-add computation of correlation matrix for every data set that it is input in TMVA.
This should be probably changed to be an option, because in case of large dataset (many features and many events) the time spent could be too large

etejedor committed:
Refactor code in implementation of TTree __getattr__

etejedor committed:
Document the TTree pythonization function

etejedor committed:
Delete patches: to be contributed upstream to Cppyy

etejedor committed:
Use a decorator for pythonizor functions

As suggested by Stefan Wunsch, decorators provide a nicer syntax
to mark that a function implements some pythonization.

etejedor committed:
Follow PEP 8 style guide for indentation

etejedor committed:
Add documentation for pythonization decorator

Axel-Naumann committed:
[cling] Move NullDerefProtectionTransformer into unnamed namesp.

steremma committed:
Simplified the APIs and reduced code duplication in `ConvLayer` and `MaxPoolingLayer`

* Pooling is now a subclass of Convolutional Layer. As a result common functions and fields are not replicated.
* Constructor arguments that can be internally computed are eliminated.

steremma committed:
Adapted `DeepNet` to the new API

steremma committed:
Renamed kernel/filter fields to use the prefix "filter" rather than "frame". This is important to have the same naming convention everywhere.

eguiraud committed:
[DF][NFC] Order includes in RDFNodes.hxx

eguiraud committed:
[DF] Use R__ASSERT instead of cassert

eguiraud committed:
[DF] Use std::numeric_limits instead of climits

eguiraud committed:
[DF] Use std::stack instead of std::vector as a stack

It makes intent clearer, with no performance issues even if
the underlying container is a std::deque since we will never
loop over the contents of these stacks.

steremma committed:
Code review corrections

eguiraud committed:
[DF] Use stacks rather than single output files in Snapshot MT

This resolves ROOT-9456, "[Jenkins][DF] Sporadic failures in test_snapshotNFiles".

The issue is due to interleaved TBB task execution in the same thread.

When nested task spawning is present (i.e. when TBB tasks spawn other
TBB tasks and the thread pool contains at least three threads), the TBB
task stealing mechanism makes it possible that a thread starts executing
a new instance of the "parent" tasks before it finishes the execution of
the previous parent task (this happens when the thread is waiting for a
"stolen" child task to finish, so it steals another threads' parent
task).

In the case of a multi-thread snapshot, one needs to write at least two
branches for the multi-thread branch writing to kick in and have nested
task spawning.

The crash was caused by a new parent task overwriting the output TTree
of the previous parent task in a given thread. When the previous parent
task resumed execution, it ended up trying to flush out the contents of
the deleted output TTree.

The solution adopted is to use thread-local stacks of output TTrees (as
opposed to single output TTrees), that tasks push and pop when they
start and finish respectively.

dpiparo committed:
[DF] TColumnValue: replace constexpr static data member with integral constant

in order to simplify this code.

Axel-Naumann committed:
[cling] Only check pointers for Decls from writable dirs (ROOT-9377).

eguiraud committed:
[DF] Expect test death rather than asserting it

The gtest can safely go on in case this test case fails.

dpiparo committed:
[DF] Due to a bug in clang, disable extern TColumnValue templates for c++17

see ROOT-9499.

dpiparo committed:
[DF] Add TColumnValue<T> autoload keys to avoid undefined symbols

on some platforms.

eguiraud committed:
[DF][NFC] Remove obsolete comments

eguiraud committed:
[DF][NFC] Remove redundant forward declarations

eguiraud committed:
[DF] Add two helper aliases for TypeList operations

eguiraud committed:
[DF] More readable tag dispatching in RCustomColumn

dpiparo committed:
[NFC] Formatting

eguiraud committed:
[DF] More readable tag dispatching for actions

eguiraud committed:
[DF] Avoid TChain::AddClone in Snapshot when unnecessary

Since output trees are recreated for each task, and each task only
processes one tree cluster (and never crosses file boundaries), we
don't need to add the output trees as clones of the input tress
unless the input trees have friends (which might cross file boundaries)
even if the main tree does not.

Usage of AddClone here is undesirable in the general case because
it generates many (safe) warnings printed at screen during the event
loop, see ROOT-9487.

eguiraud committed:
[DF] Remove obsolete data members and methods from RCustomColumnBase

They were added when custom columns where nodes of the computation
graph, but are useless now that they are not anymore.

eguiraud committed:
[DF] Add RJittedCustomColumn, in line with RJittedFilter

eguiraud committed:
[DF] Use RJittedCustomColumn when jitting custom columns

In line with what it's done for jitted filters, now the jitting
of custom columns creates the RCustomColumn and assigns it as the
fConcreteCustomColumn of the corresponding RJittedCustomColumn that
was previously booked.
This avoids having the situation in which a certain custom column
has been "booked" but is not yet present in the map of custom columns.

It will also help with the "local custom columns" that are coming
with ROOT-9465.

eguiraud committed:
[DF] Help windows with friend declaration of printValue

Axel-Naumann committed:
CMake: error out on afs glite sapdb srp.

Axel-Naumann committed:
CMake: error out for chirp or ios; they are not CMake-d.

Axel-Naumann committed:
Add thisroot.fish for the fish shell!

eguiraud committed:
[DF] Change TSlotStackPutBackTooMany to only test the nominal case

dpiparo committed:
[Core] Split critical section in TClingClassInfo::GetBaseOffset

the full method was protected with the global lock. Now we split it
in two parts, the first one where the cache is looked at and no state
is modified with a read lock. The second, where the cache is filled
and the interpreter is invoked with a write lock.

eguiraud committed:
[DF] Use numeric_limits::max instead of UINT_MAX (part 2)

Axel-Naumann committed:
[cling][NFC] Add GCC fallthrough comment.

Axel-Naumann committed:
[cling][NFC] Add GCC fallthrough comment, 80cols.

dpiparo committed:
[DF] TSlotStack uses local maps indexed with thread ids rather than thread local

this prevents bogus behaviour if two instances of TSlotStack are used simultaneously
or even one after the other.

Axel-Naumann committed:
[RelNotes] Mention packages removed for v6.16.

eguiraud committed:
[DF][NFC] Rename JitActions -> BuildJittedNodes

We jit actions as well as Define's and Filter's now.

couet committed:
When the histograms' title's font was set in pixel the position of the
`TPaveText` containing the title was not correct.

gganis committed:
netxng: add possibility to control log verbosity with NetXNG.Debug

Addresses ROOT-9311

Axel-Naumann committed:
[cling] Test printValue on unutterable types.

Axel-Naumann committed:
[cling] Check file is C_User before hitting disk. Thanks, Vassil!

Axel-Naumann committed:
Remove more traces of ios and chirp.

Axel-Naumann committed:
Mention thisroot.fish.

dpiparo committed:
[CMake] Turn on ccache by default on 32bits builds

eguiraud committed:
[DF] Remove obsolete part from error message

ashlaban committed:
[TMVA] -- Avoid unsafe reinterpret_cast

pcanal committed:
Warn about problematic base class StreamerElement only once

pcanal committed:
Insure we call SetRead/FillSequence after sub-branches have been added.

This is need to be able to distinguish split node from unsplit node.

pcanal committed:
Properly set fOnfileObject for the parent

dpiparo committed:
[DF] Do not compare type infos but rather their names

in order to take into account differences arising when one of the two
typeinfos come from a jitted context.

Axel-Naumann committed:
[rootrc] Replace Root.ZipMode by Root.CompressionAlgorithm, passing it to R__SetZipMode().

Axel-Naumann committed:
[thisroot.fish] Clean up env vars; fix `cd -`.

couet committed:
Change v7 (graphics) classes' names from Txxx to Rxxx (#2200)

* Change v7 (graphics) classes' names from Txxx to Rxxx

* more changes in linkdef.h

* Change the test macros.

* Modify v7 tutorials

* missed a "TPalette"

* T->R translate more tests.

* Update globbing to look for ROOT/R*.h

* webdisplay and fitpanelv7 are not yet converted to "R" naming.

* Fix typo

* Fix typo
increase space between lines

SimeonEhrig committed:
Fix doxygen error.

ashlaban committed:
[TMVA] Minimization-DNN remove input randomness (#2251)

Sporadic test failures are caused by the randomness of the test input.
(For some inputs the training does not converge). This patch "fixes"
this by removing the randomness from the input to the tests.

One source of randomness still remains, the dataloader shuffles the
batches internally, relying on a source of randomness that is not
reachable from the outside.

However, the variability in test output is _significantly_ reduced.

gganis committed:
netxng: give priority to Xrd_LOGLEVEL over NetXNG.Debug

amadio committed:
Do not exclude etc/http directory from installation

This was an unintended side-effect of a previous commit:
9b4d0d8dccbd48d21b0f7c79b6eaf94428f691b1.

pcanal committed:
Remove obsolete (commented-out) code fragement

pcanal committed:
Factor out ActionSequence creation

pcanal committed:
Apply new ActionSequence creation to Fill

pcanal committed:
Remove obsolete TBranchElement::fIDs

ellert committed:
Fix test compilation on RHEL 7

eguiraud committed:
Update RDF code owners

ravikiran0606 committed:
[TMVA] API-Support for SGD Optimizer:

* Implement Base Class VOptimizer

* Implement Class TSGD with Momentum method

* Add Unit Tests for TSGD

* Add Parsing Options for Optimizers

* Modify MethodDL to include the optimizer TSGD

* Add ROOT Style docs and clang-format the code

stwunsch commented on tmva/tmva/inc/TMVA/DNN/Functions.h:
```diff
@@ -76,6 +76,17 @@ enum class EInitialization {
+/* Enum representing the optimizer used for training. */
+enum class EOptimizer {
+   kSGD      = 0,
+   kAdagrad  = 1,
```
Remove the enums, which are not yet there.

stwunsch commented on tmva/tmva/inc/TMVA/DNN/Optimizer.h:
```diff
@@ -0,0 +1,100 @@
+      inline Layer_t* GetLayerAt(size_t i) { return fDeepNet.GetLayerAt(i); }
+
+      /*! Setters */
+      inline void SetGlobalStep(size_t x) { fGlobalStep = x; }
```
This is needed at any point?

stwunsch commented on tmva/tmva/inc/TMVA/DNN/Optimizer.h:
```diff
@@ -0,0 +1,100 @@
+template <typename Architecture_t, typename Layer_t, typename DeepNet_t>
+VOptimizer<Architecture_t, Layer_t, DeepNet_t>::VOptimizer() : fLearningRate(0),fGlobalStep(0)
+{
+   // nothing to do here..
```
remove the comment

stwunsch commented on tmva/tmva/inc/TMVA/DNN/Optimizer.h:
```diff
@@ -0,0 +1,100 @@
+template <typename Architecture_t, typename Layer_t, typename DeepNet_t>
+VOptimizer<Architecture_t, Layer_t, DeepNet_t>::VOptimizer(Scalar_t learningRate, DeepNet_t &deepNet) : fLearningRate(learningRate),fGlobalStep(0),fDeepNet(deepNet)
+{
+   // nothing to do here..
```
remove the comment

stwunsch commented on tmva/tmva/inc/TMVA/DNN/Optimizer.h:
```diff
@@ -0,0 +1,100 @@
+template <typename Architecture_t, typename Layer_t, typename DeepNet_t>
+VOptimizer<Architecture_t, Layer_t, DeepNet_t>::~VOptimizer()
+{
+   // nothing to do here..
```
remove the comment

stwunsch commented on tmva/tmva/inc/TMVA/DNN/Optimizer.h:
```diff
@@ -0,0 +1,100 @@
+ * Authors (alphabetical):                                                        *
+ *      Ravi Kiran S      <sravikiran0606@gmail.com>  - CERN, Switzerland         *
+ *                                                                                *
+ * Copyright (c) 2005-2015:                                                       *
```
Make the header correct in all files (see copyright, ...)

stwunsch commented on tmva/tmva/inc/TMVA/DNN/Optimizer.h:
```diff
@@ -0,0 +1,100 @@
+#ifndef TMVA_DNN_OPTIMIZER
+#define TMVA_DNN_OPTIMIZER
+
+#include <iostream>
```
needed?

steremma committed:
Fix warnings introduced by #2229

stwunsch commented on tmva/tmva/inc/TMVA/DNN/SGD.h:
```diff
@@ -0,0 +1,166 @@
+template <typename Architecture_t, typename Layer_t, typename DeepNet_t>
+TSGD<Architecture_t, Layer_t, DeepNet_t>::TSGD()
+{
+   // nothing to do here..
```
remove

stwunsch commented on tmva/tmva/inc/TMVA/DNN/SGD.h:
```diff
@@ -0,0 +1,166 @@
+   fPastBiasGradients.resize(layersNSlices);
+
+   for (size_t i = 0; i < layersNSlices; i++) {
+      size_t weightsNSlices = (deepNet.GetLayerAt(i)->GetWeights()).size();
```
make const (?)

stwunsch replied:
or use auto

stwunsch commented on tmva/tmva/inc/TMVA/DNN/SGD.h:
```diff
@@ -0,0 +1,166 @@
+   for (size_t i = 0; i < layersNSlices; i++) {
+      size_t weightsNSlices = (deepNet.GetLayerAt(i)->GetWeights()).size();
+      size_t weightsNRows = 0;
+      size_t weightsNCols = 0;
```
move these to the block of the for loop and make const

stwunsch commented on tmva/tmva/inc/TMVA/DNN/SGD.h:
```diff
@@ -0,0 +1,166 @@
+         initialize<Architecture_t>(fPastWeightGradients[i][j], EInitialization::kZero);
+      }
+
+      size_t biasesNSlices = (deepNet.GetLayerAt(i)->GetBiases()).size();
```
same here as for the weights (move to block and try to optimize)

stwunsch commented on tmva/tmva/inc/TMVA/DNN/SGD.h:
```diff
@@ -0,0 +1,166 @@
+template <typename Architecture_t, typename Layer_t, typename DeepNet_t>
+TSGD<Architecture_t, Layer_t, DeepNet_t>::~TSGD()
+{
+   // nothing to do here..
```
remove

stwunsch commented on tmva/tmva/inc/TMVA/DNN/SGD.h:
```diff
@@ -0,0 +1,166 @@
+
+//_________________________________________________________________________________________________
+template <typename Architecture_t, typename Layer_t, typename DeepNet_t>
+void TSGD<Architecture_t, Layer_t, DeepNet_t>::Print() const
```
what's the use of this method?

stwunsch commented on tmva/tmva/src/MethodDL.cxx:
```diff
@@ -342,6 +344,23 @@ void MethodDL::ProcessOptions()
       }
 
+      TString optimizer = fetchValueTmp(block, "Optimizer", TString("SGD"));
+      if (optimizer == "SGD") {
```
put only optimizers here, which are implemented (SGD)

stwunsch commented on tmva/tmva/src/MethodDL.cxx:
```diff
@@ -1092,15 +1113,50 @@ void MethodDL::TrainDeepNet()
+      DNN::TSGD<Architecture_t, Layer_t, DeepNet_t> minimizer(settings.learningRate, deepNet, settings.momentum, kFALSE);
+          
+
+      // switch(O) {
```
remove outcommented code

stwunsch commented on tmva/tmva/src/MethodDL.cxx:
```diff
@@ -1138,7 +1194,8 @@ void MethodDL::TrainDeepNet()
 
       while (!converged) {
-         stepCount++;
+         //stepCount++;
```
dont't outcomment code

stwunsch commented on tmva/tmva/src/MethodDL.cxx:
```diff
@@ -1154,25 +1211,26 @@ void MethodDL::TrainDeepNet()
-               //minimizer.Step(deepNet, nets, batches);
-               minimizer.Step(deepNet, my_batch.GetInput(), my_batch.GetOutput(), my_batch.GetWeights());
-            }
+            // execute one minimization step
```
don't outcomment code

stwunsch commented on tmva/tmva/test/DNN/TestOptimization.h:
```diff
@@ -0,0 +1,290 @@
+   fillMatrix(WTest, 1.0);
+
+   // Construct the deepNet
+   size_t inputDepth = 1;
```
you could make all of these const (?) just details, does not matter too much

stwunsch commented on tmva/tmva/test/DNN/TestOptimization.h:
```diff
@@ -0,0 +1,290 @@
+
+
+   // Initialize the optimizer
+   Optimizer_t minimizer(0.001,deepNet,momentum,false);
```
why is the optimizer called minimizer? this is kind of legacy naming which you should get rid of

stwunsch commented on tmva/tmva/src/MethodDL.cxx:
```diff
@@ -1092,15 +1113,50 @@ void MethodDL::TrainDeepNet()
-                                                        settings.testInterval);
+      // DNN::TDLGradientDescent<Architecture_t> minimizer(settings.learningRate, settings.convergenceSteps, settings.testInterval);
+
+      DNN::TSGD<Architecture_t, Layer_t, DeepNet_t> minimizer(settings.learningRate, deepNet, settings.momentum, kFALSE);
```
call the optimizer `optimizer` and not minimizer. this is confusing and only because of the legacy code. no need to do this!

stwunsch reviewed (changes requested):
Alright, looks good! I made a lot of comments, but nothing crucial.

stwunsch commented:
and the test does not seem to have clang-format formatting (?)

ellert committed:
Remove remaining references to removed Java configuration variables

ellert committed:
Add missing include for std::shuffle

xvallspl committed:
fix wrong early return in hadd

imaxoi committed:
[DF] Introduced Standard Deviation Action

imaxoi committed:
[DF] Fixed test to work in multithread

shahor02 committed:
Fix in addition of template fwd declaration in rootmap

couet committed:
Simplify example

couet committed:
The error on angle position along the radius doesn't depend on the angle's units (rad, deg, grad).

Fix; https://sft.its.cern.ch/jira/browse/ROOT-8476

etcherniaev committed:
Get rid of the limitation for PHI sectors (#2265)

ellert committed:
Adjust reference for Aarch64 (#2120)

stwunsch commented on tmva/tmva/inc/TMVA/DNN/Optimizer.h:
```diff
@@ -0,0 +1,100 @@
+      virtual void UpdateBiases(std::vector<Matrix_t> &biases,const std::vector<Matrix_t> &biasGradients) = 0;
+
+   public:
+      VOptimizer();
```
Remove default constructor (?)

stwunsch commented on tmva/tmva/inc/TMVA/DNN/SGD.h:
```diff
@@ -0,0 +1,166 @@
+template <typename Architecture_t, typename Layer_t, typename DeepNet_t>
+TSGD<Architecture_t, Layer_t, DeepNet_t>::TSGD(Scalar_t learningRate, DeepNet_t &deepNet, Scalar_t momentum, Bool_t isNesterov) : VOptimizer<Architecture_t, Layer_t, DeepNet_t>(learningRate,deepNet),fMomentum(momentum),fIsNesterov(isNesterov)
+{
+   size_t layersNSlices = deepNet.GetLayers().size();
```
use auto

imaxoi committed:
[DF] Std deviation unbiased

Solves ROOT-9460

imaxoi committed:
[DF] Renaming, tests on cornercases

imaxoi committed:
[DF] Added gtest for StdDev precision

dpiparo committed:
[DF] Add cache examples

dpiparo committed:
[DF][NFC] Make clang format happy

imaxoi committed:
[DF] Added test for StdDev on collections

imaxoi committed:
Changed tests to be executed in multithread

imaxoi committed:
Improved precision test

imaxoi committed:
[DF] Tests using EXPECT_DOUBLE_EQ instead of EXPECT_NEAR

bellenot committed:
Fix line width and maker index issues on Windows

dpiparo committed:
[DF] Document better caching tutorials

eguiraud committed:
[DOCS] Update RDF release notes

linev committed:
xml: when reallocate buffer, also optionally recalculate current
position

linev committed:
xml: use only offset from current position in LocateValue

If current pointer changes due to reallocation offset can be used as is

linev committed:
xml: fix in bugfix

bellenot committed:
Implement splash screen version 6 on Windows

stwunsch commented on tmva/tmva/inc/TMVA/DNN/SGD.h:
```diff
@@ -41,76 +40,77 @@ namespace DNN {
+      fPastBiasGradients; ///< The sum of the past bias gradients associated with the deep net.
+
+   /*! Update the weights, given the current weight gradients. */
+   void UpdateWeights(std::vector<Matrix_t> &weights, const std::vector<Matrix_t> &weightGradients);
```
Is it needed to have two different functions to update weights and biases?

stwunsch commented on tmva/tmva/inc/TMVA/DNN/Optimizer.h:
```diff
@@ -0,0 +1,101 @@
+   DeepNet_t &fDeepNet;    ///< The reference to the deep net.
+
+   /*! Update the weights, given the current weight gradients. */
+   virtual void UpdateWeights(std::vector<Matrix_t> &weights, const std::vector<Matrix_t> &weightGradients) = 0;
```
Proposal: Remove the `UpdateWeights` and `UpdateBiases` and make a helper function `void Update(std::vector weights, std::vector gradients)`, which simply ralls the `ScaleAdd` on all elements.

stwunsch commented on tmva/tmva/inc/TMVA/DNN/Optimizer.h:
```diff
@@ -0,0 +1,101 @@
+   VOptimizer(Scalar_t learningRate, DeepNet_t &deepNet);
+
+   /*! Performs one step of optimization. */
+   virtual void Step() = 0;
```
Implement `Step` in the base class.

pcanal committed:
In GatherArtificialElements do not drill through transient data member.

They are guaranteed to not have induced any branches and often do not have dictionary information (or have members/bases that do not).

Fixes ROOT-9513
And see https://github.com/cms-sw/cmssw/pull/22594

linev committed:
xml: use variable reference in ExpandStream

Introduce dummy for it

linev committed:
xml: no need to reassign curr variable - not a dummy now

eguiraud committed:
[DOCS] Add me and Massimo to contributor list

agheata committed:
Less verbose output when reading GDML auxiliary info.

agheata committed:
Added protection for ComputeLradTsaiFactor when Z=0

agheata committed:
Fix for ROOT-7917

xvallspl committed:
Fix overallocating in the chunked mapping

xvallspl committed:
Remove non thread safe additions

Remove non thread safe additions when parallelizing the fit

eguiraud committed:
[DF] Add test for validation of custom column names

eguiraud committed:
[DF] Throw if name of defined column is not a valid C++ var name

amadio committed:
Add ROOTDataFrame to ${ROOT_LIBRARIES}

eguiraud committed:
[DF][NFC] Minor fixes to RDFHelpers.hxx

- moved internals from ROOT::Internal to ROOT::Internal::RDF
- removed useless includes

eguiraud committed:
[DF][NFC] Doxygen doc formatting

eguiraud committed:
[DF] Add PassAsVec helper function

PassAsVec<N, T>(func) is a callable that takes N arguments of type T,
passes them to func as a collection (`func({v1, v2, ...}`) and returns
the result of the call to `func`.

This helper makes it possible to pass several columns of the same
type to a callable that accepts a vector of that type.

eguiraud committed:
[DF] Minor improvements to dataframe_helpers test

eguiraud committed:
[DF] Add a test for the PassAsVec RDF helper

xvallspl committed:
Generalize vectorized Binned ML code

linev committed:
jsroot 5.5.0

steremma committed:
fix warnings

Axel-Naumann committed:
[v7hist] Make histograms more regular (copyable values).

whit2333 committed:
Updated TDataSource documentation

peremato committed:
Added patch to ROOT for gcc 8.1

peremato committed:
Revert "Added patch to ROOT for gcc 8.1"

This reverts commit bf2cc576ae7f91861b83a0deb0ec81e445aff15f.

Axel-Naumann committed:
[v7hist] Rename THist->RHist. Enumerate files.

steremma committed:
Verbose argument set for forward convolution methods in all architectures

steremma committed:
Compatibility with TCpuMatrix

steremma committed:
Helper functions for propagation and testing in both architectures (reference not supported)

steremma committed:
adapted CPU propagation to the new argument set

steremma committed:
Convolutional Layer implemention for the GPU

steremma committed:
Added tests for all architectures for the RotateWeights methods

steremma committed:
Added test cases for forward and backward propagation through a single convolution layer in both architectures.

I had to revert the debugging mechanism to use the lower level `Print()` because `PrintMatrix` only works for the CPU case.

steremma committed:
Ported the integration test (finite difference or gradient check) into CUDA

steremma committed:
registered all new tests in CMakeLists

steremma committed:
fixed shadowed declaration warning

eguiraud committed:
[DF] Add tutorial on RDF helpers

steremma committed:
Fixed warnings

Axel-Naumann committed:
[v7hist] Make STORAGE a type template arg, remove from STAT.

Simplifies types, considerably.

eguiraud committed:
[DF] Make RResultPtr members non-const to allow copy/move assignment

Fixes ROOT-9526.

eguiraud committed:
[DF] Make RResultPtr move-constructible, add a test for it

eguiraud committed:
[DF] Fix warnings relative to PassAsVecHelper

dpiparo committed:
[TMVA] Remove warnings

dpiparo committed:
[TMVA] Remove another warning about unused variable

Axel-Naumann committed:
Make TBits::op~() const-correct [ROOT-9529].

Axel-Naumann committed:
TBits: remove __CINT__ remnants.

yamaguchi1024 committed:
Fix Wsign-compare warning in master

couet committed:
Fix in GetHistogram()

In TGraph2D::GetHistogram when the points were all in the same plane (along X or Y) at a
negative coordinate, the computed axis limits were not correct. This was reported
here: https://root-forum.cern.ch/t/potential-bug-in-tgraph2d/29700/5.

FonsRademakers committed:
Restore FitSlices() call back to original parameters.

Shows again same results as in the tutorials/dataframe/df101_h1Analysis.C
script.

bellenot committed:
Add support for Python 3 on Windows

This fixes the issue reported on the forum at: https://root-forum.cern.ch/t/windows-10-source-installation-failing-due-to-python-3-bindings/29771
